booster/library/Booster/SMT/Interface.hs

```diff
@@ -156,28 +150,88 @@ translatePrelude def =
 
 checkPrelude :: Log.LoggerMIO io => SMT io ()
 checkPrelude = do
-    runCmd_ $ SetTimeout defaultSMTOptions.timeout
+    -- set large default timeout value for checking the prelude
+    setTimeout defaultSMTOptions.timeout
     Log.logMessage ("Checking definition prelude" :: Text)
-    check <- runPrelude >> runCmd CheckSat
+    ctxt <- SMT get
+    -- \| Send the commands from the definition's SMT prelude
+    check <- mapM_ runCmd ctxt.prelude >> runCmd CheckSat
+    -- set user defined timeout value for the general queries
+    setTimeout ctxt.options.timeout
     case check of
         Sat -> pure ()
         other -> do
             Log.logMessage $ "Initial SMT definition check returned " <> pack (show other)
-            SMT get >>= closeContext
+            closeContext ctxt
             throwSMT' $
                 "Aborting due to potentially-inconsistent SMT setup: Initial check returned " <> show other
-
--- | Send the commands from the definition's SMT prelude
-runPrelude :: Log.LoggerMIO io => SMT io ()
-runPrelude = do
-    prelude <- SMT $ gets prelude
-    mapM_ runCmd prelude
+  where
+    setTimeout timeout = do
+        Log.logMessage $ "Setting SMT timeout to: " <> show timeout
+        runCmd_ $ SetTimeout timeout
 
 finaliseSolver :: Log.LoggerMIO io => SMT.SMTContext -> io ()
 finaliseSolver ctxt = do
     Log.logMessage ("Closing SMT solver" :: Text)
     destroyContext ctxt
 
+data IsSatResult unknown a = IsSat a | IsUnsat | IsUnknown unknown
```

Comment:
Let's add another datatype here:

```
data IsValidResult unknown a = IsValid a | IsInvalid | IsUnknown unknown
```

and use it as the return type of `checkPredicates`.

Reply:
how about making them pattern synonyms so that we can reuse `IsUnknown`? i.e.

```haskell
type IsValidResult = IsSatResult (Maybe Text) ()

pattern IsValid, IsInvalid :: IsValidResult
pattern IsValid = IsSat ()
pattern IsInvalid = IsUnsat

{-# COMPLETE IsValid, IsInvalid, IsUnknown #-}
```

Reply:
I'd still prefer a separate datatype, but this is an acceptable option too. @jberthold do you have a strong opinion?